geps/gep-1767/index.md

```diff
@@ -0,0 +1,701 @@
+
+The optional response header `Access-Control-Expose-Headers` controls which HTTP response headers are exposed to clients for the actual cross-origin request. 
+
+If the server specifies the response header `Access-Control-Allow-Credentials: true`, the actual cross-origin request will be able to use credentials for getting sensitive resources. 
+Credentials are cookies, TLS client certificates, or authentication headers containing a username and password.
+
+After the server has permitted the CORS "preflight" request, the client will be able to send actual cross-origin request.
+If the server doesn’t want to allow cross-origin access, it will respond with an error message to the client.
+
+## API
+This GEP proposes to add a new field `HTTPCORSFilter` to `HTTPRouteFilter`.
+If HTTPCORSFilter is set, then the gateway will generate the response of the preflight requests and send back it to the client directly.
+For the actual cross-origin request, the gateway will add CORS headers to the response before it is sent to the client.
+
+```golang
+const (
+    // HTTPRouteFilterCORS can be used to add CORS headers to an 
+    // HTTP response before it is sent to the client.
+    //
+    // Support in HTTPRouteRule: Extended
+    //
+    // Support in HTTPBackendRef: Extended
+    HTTPRouteFilterCORS HTTPRouteFilterType = "CORS"
+)
+
+type HTTPRouteFilter struct {
+    // CORS defines a schema for a filter that responds to the
+    // cross-origin request based on HTTP response header.
+    // 
+    // Support: Extended
+    //
+    // +optional
+    CORS *HTTPCORSFilter `json:"cors,omitempty"`
+}
+
+// HTTPCORSFilter defines a filter that responds to the
+// cross-origin request based on HTTP header.
+type HTTPCORSFilter struct {
+    // AllowOrigins indicates whether the response can be shared with 
+    // requested resource from the given `Origin`.
+    // 
+    // A wildcard indicates that the requests from all `Origin` are allowed.
+    //
+    // When responding to a credentialed requests, the gateway must specify 
+    // an origin in the value of the Access-Control-Allow-Origin response header, 
+    // instead of specifying the * wildcard.
+    //
+    // Input:
+    //   Origin: https://foo.example
+    //
+    // Config:
+    //   allowOrigins: ["*"]
+    //
+    // Output:
+    //   Access-Control-Allow-Origin: *
+    //
+    // For a simple cross-origin interaction, the gateway sets value of the header 
+    // Access-Control-Allow-Origin same as the `Origin` header provided by the client.
+    //
+    // The status code of a successful response to a "preflight" request is an OK status (i.e., 204 or 200). 
+    // For the "preflight" request, if the request `Origin` dos not match the configured allowed origins,
+    // the gateway will return a response with error status (e.g., 403). 
```

Comment:
I don't think this is how CORS works generally. Or, at the very least, not how Envoy's CORS filter works which is what I am familiar with so I will desribe that.

In envoy, if a request is a CORS request then there will be a direct response from the CORS filter. A "CORS request" means:

- method is OPTIONS
- "origin" header is not empty
- "Access-Control-Request-Method" is not empty"

If it meets this criteria, it will always return a 200 with the various CORS response headers. If its not allowed, it will be rejected by the browser -- NOT the server.

If it doesn't meet the criteria, the request is just passed through